File: 0877-stone-game/0877-stone-game.java
class Solution {
    int[][][] dp;
    public boolean stoneGame(int[] piles) {
        int n = piles.length;
        int[][][] dp = new int[2][n][n];
        for(int[][] arr:dp) {
            for(int[] sub:arr) Arrays.fill(sub, -1);
        }
        return solve(piles, dp, 0, n-1, 1) > 0;
    }
    int solve(int[] piles, int[][][] dp, int i, int j, int person) {
        if(i > j) return 0;
        if(dp[person][i][j] != -1) return dp[person][i][j];
        if(person == 1) {
            // Alice will maximize the score
            return dp[person][i][j] = Math.max(piles[i] - solve(piles, dp, i+1, j, 0), piles[j] - solve(piles, dp, i, j-1, 0));
        }
        else {
            // Bob will minimize the score
            return dp[person][i][j] = Math.min(-piles[i] - solve(piles, dp, i+1, j, 1), -piles[j] - solve(piles, dp, i, j-1, 1));
        }
    }
}

File: 0877-stone-game/0877-stone-game.cpp
class Solution {
public:
    int dp[2][501][501];
    bool stoneGame(vector<int>& piles) {
        memset(dp, -1, sizeof(dp));
        int n = piles.size();
        return solve(piles, 0, n-1, 1) > 0;
    }
    int solve(vector<int>&piles, int i, int j, int person) {
        if(i > j) return 0;
        if(dp[person][i][j] != -1) return dp[person][i][j];
        if(person == 1) {
            // Alice should maximize the score
            dp[person][i][j] = max(piles[i] - solve(piles, i+1, j, 0), piles[j] - solve(piles, i, j-1, 0));
        }
        else {
            dp[person][i][j] = min(-piles[i] - solve(piles, i+1, j, 1), -piles[j] - solve(piles, i, j-1, 1));
        }
        return dp[person][i][j];
    }
};